Sources/NIOCore/NIOLoopBound.swift

```diff
@@ -0,0 +1,120 @@
+    @inlinable
+    public init(_ value: Value, eventLoop: EventLoop) {
+        eventLoop.preconditionInEventLoop()
+        self._eventLoop = eventLoop
+        self._value = value
+    }
+
+    /// Access the `value` with the precondition that the code is running on `eventLoop`.
+    ///
+    /// - note: ``NIOLoopBound`` itself is value-typed, so any writes will only affect the current value.
+    @inlinable
+    public var value: Value {
+        get {
+            self._eventLoop.preconditionInEventLoop()
+            return self._value
+        }
+        set {
+            self._eventLoop.preconditionInEventLoop()
+            self._value = newValue
+        }
+    }
+}
+
+/// ``NIOLoopBoundBox`` is an always-``Sendable``, reference-typed container allowing you access to ``Value`` if and
+/// only if on the right EventLoop``.
+///
+/// ``NIOLoopBoundBox`` is useful to transport a value of a non-``Sendable`` type that needs to go from one place in
+/// your code to another where you (but not the compiler) know is on one and the same ``EventLoop``. Usually this
+/// involves `@Sendable` closures. This type is safe because it verfies (using `eventLoop.preconditionInEventLoop()`)
+/// that this is actually true.
+///
+/// A ``NIOLoopBoundBox`` can only be read from or written to when you a provably
+/// (through `eventLoop.preconditionInEventLoop()`) on the ``EventLoop`` associated with the ``NIOLoopBound``. Accessing
+/// or constructing it from any other place will crash your program with a precondition as it would be undefined
+/// behaviour to do so.
+///
+/// If constructing a ``NIOLoopBoundBox`` with a `value`, it is also required the program to already be on `eventLoop`
+/// but if you have a ``NIOLoopBoundBox`` that contains an ``Optional`` type, you may initialise it _without a value_
+/// whilst off the ``EventLoop``. Any read/write access to `value` afterwards will require you to be on `eventLoop`.
+public final class NIOLoopBoundBox<Value>: @unchecked Sendable {
+    public let _eventLoop: EventLoop
+
+    @usableFromInline
+    /* private */var _value: Value
+
+    /// Initialise a ``NIOLoopBoundBox`` to `value` with the precondition that the code is running on `eventLoop`.
+    @inlinable
+    public init(_ value: Value, eventLoop: EventLoop) {
+        // This precondition is absolutely required. If not, it were possible to take a non-Sendable `Value` from
+        // _off_ the ``EventLoop`` and transport it _to_ the ``EventLoop``. That would be illegal.
+        eventLoop.preconditionInEventLoop()
+        self._eventLoop = eventLoop
+        self._value = value
+    }
+
+    /// Initialise a ``NIOLoopBoundBox`` with `nil`, this does _not_ require you to be running on `eventLoop`.
+    public init<NonOptionalValue>(
+        valueType: NonOptionalValue.Type = NonOptionalValue.self,
+        eventLoop: EventLoop
+    ) where Optional<NonOptionalValue> == Value {
```

Comment:
Usability nit: I'm not sure `NIOLoopBound(valueType: Foo.self, eventLoop: loop)` is not immediately obvious that it's for an optional value.

The other small usability concern is that you can also do `NIOLoopBoundBox<Foo?>(nil, eventLoop: loop)` off of the event loop and still crash when `NIOLoopBound(valueType: Foo.self, eventLoop: loop)` wouldn't.

I wonder if instead of this `init` we should overload the other init:

```swift
public init<NonOptionalType>(_ value: NonOptionalType?, eventLoop: EventLoop) where Optional<NonOptionalType> == Value {
    if value == nil { 
        // all good 
    } else { 
        // check on EL.
    }
}
```

We may have to mark the other `init` as `dis

Reply:
@glbrntt What about we spell the `nil` `init` one as a `static func makeEmpty()`?

I think it's confusing to have some `init`s okay and others not if you're off the EL. And the other issue with `init`s is that we can't easily add extra words like "empty" or something.

My new suggestion:

- all `init`s (there'll be just one) require to be called _on_ the EL
- there's a static factory that can be called off the EL which initialises an empty one.

WDYT? Also CC @Lukasa & @dnadoba

Reply:
implementation suggestion here: https://github.com/apple/swift-nio/pull/2370/commits/c58431796c5b481df61368ab579cd43c4dd8ea55

Reply:
(also added to this PR as a separate commit)

Reply:
"Empty" isn't something I would associated with an Optional but rather with a Collection. Maybe we call it `makeNone()` or even just `.none`?
(We also have `ExpressibleByNilLiteral` which would allow us to use the `nil` literal [but the documentation discouraged its usage](https://developer.apple.com/documentation/swift/expressiblebynilliteral) and we couldn't specify the event loop which we must in our case.)

Reply:
`makeNone` sounds kinda weird to me. `makeBoxWithNil`? Don't think `nil` literal would work because we need the EL